[4.1] Making Google happy about breadcrumbs again.

crystalenka opened the pull request:
This PR fixes structured data errors that made Google flag breadcrumbs as invalid.

When a breadcrumb item was a menu heading, instead of a link, Google would flag it as an invalid breadcrumb because it did not have a link. This is fixed by removing the JSON markup for items without a link, with the exception of the last item/current page.

This pull request fixes issues #37508 and #37510.

Basically reproducing my changes from my prior 4.0 pull request, but updated for the new JSON format.

### Summary of Changes

1. Added a check if a link exists in a breadcrumb item. If not, skip it and don't add it to the JSON.
2. Added the last item (aka current page) to the JSON without a link. This is valid per [Google SEO guidelines](https://developers.google.com/search/docs/advanced/structured-data/breadcrumb#list-item); they will pull the current page URL for the last item.

### Testing Instructions

Create a site menu with a menu heading and subpages, like so:

- Home
- Menu Heading
  - Subpage 1

Create a breadcrumbs module. Ensure the module is published and visible on Subpage 1.

Navigate to subpage 1.

If this site is accessible over the internet:

Copy the URL and paste it into the field here: https://search.google.com/test/rich-results

If this site is hosted locally:

View source of Subpage 1 and copy it into the code box on https://search.google.com/test/rich-results
(Alternatively, you can find the generated JSON for the module and just copy that into the code box.)

### Actual result BEFORE applying this Pull Request

Google detects the breadcrumbs but shows an error complaining about the missing URL in an item.

### Expected result AFTER applying this Pull Request

Google correctly detects the breadcrumbs and marks them as valid.

### Documentation Changes Required

None as far as I am aware.

### Additional Notes

I intentionally am not checking if the module is set to display the last item or not because for the purposes of the JSON, it is better to include it and does not affect the front end display of the module.

crystalenka committed:
Making google happy with breadcrumbs again.

crystalenka committed:
Fixing current page breadcrumb JSON

Per https://developers.google.com/search/docs/advanced/structured-data/breadcrumb#list-item:

"If the breadcrumb is the last item in the breadcrumb trail, item is not required. If item isn't included for the last item, Google uses the URL of the containing page."

crystalenka committed:
Merge branch '4.1-dev' into 4.1-dev

richard67 commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -78,14 +78,27 @@
-				]
-		];
+		// Only add item to JSON if it has a valid link, otherwise skip it.
+		if(!empty($item->link)) {
```
```suggestion
		if (!empty($item->link))
		{
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

richard67 commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -78,14 +78,27 @@
+							'name' => $item->name
+					]
+			];
+		} elseif ($key === $last_item_key) {
```
```suggestion
		}
		elseif ($key === $last_item_key)
		{
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

richard67 commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -78,14 +78,27 @@
+					'item'     => [
+							'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
+							'name' => $item->name
+					]
```
```suggestion
							'name' => $item->name,
					],
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

crystalenka committed:
Apply suggestions from code review

Thank you, and apologies! I don't usually dig into the PHP; I'll double check the code style next time.

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -78,14 +78,30 @@
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'name' => $item->name
+					]
```
```suggestion
							'name' => $item->name,
					],
```
Sorry, this one I have forgotten in my last review. Of course also here for the last array elements a comma at the end due to our code style rules.

crystalenka committed:
Update modules/mod_breadcrumbs/tmpl/default.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -78,14 +78,30 @@
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
```
Does it still need the ternary operator here, now as with the if condition in line 82 `$item->link` cannot be empty at this place?

In line 56 where we have the same situation there is no ternary.

crystalenka replied:
Good point! I'll revise.

nadjak77 commented:
I have tested this item :red_circle: unsuccessfully on f034a3c4c9f705698873af506d25525d1bc00f7f<br /><br />I still get the error "Field itemListElement missing"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

crystalenka commented:
Hmmm I will take a look again, maybe I missed something. 

> On Apr 29, 2022, at 4:28 PM, nadjak77 ***@***.***> wrote:
> 
> ﻿
> I have tested this item 🔴 unsuccessfully on f034a3c
> 
> I still get the error "Field itemListElement missing"
> 
> This comment was created with the J!Tracker Application at issues.joomla.org/tracker/joomla-cms/37679.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

crystalenka committed:
Merge branch '4.1-dev' into 4.1-dev

crystalenka commented:
> I have tested this item 🔴 unsuccessfully on [f034a3c](https://github.com/joomla/joomla-cms/commit/f034a3c4c9f705698873af506d25525d1bc00f7f)I still get the error "Field itemListElement missing"

I just tested on a clean install of 4.1 with Cassiopeia template and the blog sample data. No errors for me: https://search.google.com/test/rich-results/result?id=JGXsLee4zFK1Py_gDrovhg 

@nadjak77 Can you please share your rich test results link with me by hitting the "share" button so I can debug? 

<img width="759" alt="Screen Shot 2022-05-02 at 17 02 41" src="https://user-images.githubusercontent.com/9141288/166247651-297efb6f-4222-455e-afeb-352890af4041.png">

I'm not sure how you could have gotten different results, unless maybe you have a template override on the breadcrumbs?

Thanks!

crystalenka committed:
Removing unnecessary ternary operator

nikosdion commented:
I have tested this item :red_circle: unsuccessfully on c69c9330225f1a075696b5d09ee1bdff8ada351d<br /><br />I also get the “Missing field "itemListElement"” error at https://search.google.com/test/rich-results/result/r%2Fbreadcrumbs?id=iPMy7EevFSOnJtC0n2iV5g for my test page https://nikosdiontest.joomla.com/index.php/sample-layouts/blog

However, this appears to come not from the LD-JSON microdata but from the leftover Schema.org microdata on the breadcrumbs' `<ol>` tag:

```html
<ol itemscope itemtype="https://schema.org/BreadcrumbList" class="mod-breadcrumbs breadcrumb px-3 py-2">
```

Since we're adding LD-JSON shouldn't we remove the Schema.org microdata altogether?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

crystalenka commented:
> Since we're adding LD-JSON shouldn't we remove the Schema.org microdata altogether?

Yup. I missed that, must be remnant from the original conversion to JSON instead of micro data.

crystalenka committed:
Removing leftover microdata

nikosdion commented:
I have tested this item :white_check_mark: successfully on f0f4c764b51bc9848f93164a39d4f8d30faa717f<br /><br />I can confirm that now it works as described. There are no more breadcrumb microdata errors for the same page I tested before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

richard67 commented:
@nadjak77 Could you test again with the latest changes? Thanks in advance.

richard67 commented:
@aint-broke Please test if this PR solves your issue #37510 . Thanks in advance.

nadjak77 commented:
I have tested this item :white_check_mark: successfully on f0f4c764b51bc9848f93164a39d4f8d30faa717f<br /><br />tested successfully. Thank you<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

nadjak77 commented:
@crystalenka Sorry for my late reply.
My first test (unsuccessfully) was on a clean new website without any other installations. The installation was on localhost.
But now the test was successfull. Thank you.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

crystalenka commented:
No problem, @nadjak77 !! We got it figured out!

laoneo committed:
Merge branch '4.1-dev' into 4.1-dev

laoneo commented:
Thanks!
